pkg/sql/importer/import_processor.go

```diff
@@ -394,37 +404,80 @@ func ingestKvs(
-		InitialSplitsIfUnordered: int(spec.InitialSplits),
-		WriteAtBatchTimestamp:    true,
-		ImportEpoch:              bulkAdderImportEpoch,
-	})
-	if err != nil {
-		return nil, err
+	// Setup external storage on node local for our generated SSTs.
+	var err error
+	var es cloud.ExternalStorage
+	var uriBase = fmt.Sprintf("nodelocal://%d/import/%d/", flowCtx.Cfg.NodeID.SQLInstanceID(), spec.JobID)
+	if useDistributedMerge {
+		es, err = flowCtx.Cfg.ExternalStorageFromURI(ctx, uriBase, spec.User())
+		if err != nil {
+			return nil, err
+		}
+		defer es.Close()
+	}
+
+	var pkIndexAdder kvserverbase.BulkAdder
+	var pkFileAllocator bulksst.FileAllocator
+	if !useDistributedMerge {
+		pkIndexAdder, err = flowCtx.Cfg.BulkAdder(ctx, flowCtx.Cfg.DB.KV(), writeTS, kvserverbase.BulkAdderOptions{
+			Name:                     pkAdderName,
+			DisallowShadowingBelow:   writeTS,
+			SkipDuplicates:           true,
+			MinBufferSize:            minBufferSize,
+			MaxBufferSize:            maxBufferSize,
+			InitialSplitsIfUnordered: int(spec.InitialSplits),
+			WriteAtBatchTimestamp:    true,
+			ImportEpoch:              bulkAdderImportEpoch,
+		})
+		if err != nil {
+			return nil, err
+		}
+		defer pkIndexAdder.Close(ctx)
+	} else {
+		// Setup a BulkAdder to generate SSTs into node local.
+		pkFileAllocator = bulksst.NewExternalFileAllocator(es, pkAdderName)
+		t := bulksst.NewUnsortedSSTBatcher(flowCtx.Cfg.Settings, pkFileAllocator)
+		t.SetWriteTS(writeTS)
+		pkIndexAdder = t
+		defer pkIndexAdder.Close(ctx)
 	}
-	defer pkIndexAdder.Close(ctx)
 
 	minBufferSize, maxBufferSize = importBufferConfigSizes(flowCtx.Cfg.Settings,
 		false /* isPKAdder */)
-	indexAdder, err := flowCtx.Cfg.BulkAdder(ctx, flowCtx.Cfg.DB.KV(), writeTS, kvserverbase.BulkAdderOptions{
-		Name:                     indexAdderName,
-		DisallowShadowingBelow:   writeTS,
-		SkipDuplicates:           true,
-		MinBufferSize:            minBufferSize,
-		MaxBufferSize:            maxBufferSize,
-		InitialSplitsIfUnordered: int(spec.InitialSplits),
-		WriteAtBatchTimestamp:    true,
-		ImportEpoch:              bulkAdderImportEpoch,
-	})
-	if err != nil {
-		return nil, err
+	var indexAdder kvserverbase.BulkAdder
+	var indexFileAllocator bulksst.FileAllocator
+	if !useDistributedMerge {
+		indexAdder, err = flowCtx.Cfg.BulkAdder(ctx, flowCtx.Cfg.DB.KV(), writeTS, kvserverbase.BulkAdderOptions{
+			Name:                     indexAdderName,
+			DisallowShadowingBelow:   writeTS,
+			SkipDuplicates:           true,
+			MinBufferSize:            minBufferSize,
+			MaxBufferSize:            maxBufferSize,
+			InitialSplitsIfUnordered: int(spec.InitialSplits),
+			WriteAtBatchTimestamp:    true,
+			ImportEpoch:              bulkAdderImportEpoch,
+		})
+		if err != nil {
+			return nil, err
+		}
+		defer indexAdder.Close(ctx)
+	} else {
+		// Setup a BulkAdder to generate SSTs into node local.
+		indexFileAllocator = bulksst.NewExternalFileAllocator(es, indexAdderName)
```

Comment:
I think the approach of using one writer for indexes and one writer for data works for now, but when we are tuning the performance we may want to consider using a single flusher.

This split exists for the current version of import because it index data is always out of order and primary key data might be in order. But ideally the sorting will be cheap enough that we might be better off using a single flusher since we can either reduce the amount of memory we use or use the memory more effectively.

Reply:
Fair point, let me add another TODO here for that